docs/design-docs/01-oci-support.md

```diff
@@ -0,0 +1,418 @@
+```
+
+## User Guide
+
+### Creating a package respository
+
+Before kpt packages can be pushed and pulled as OCI images, a suitable repository 
+must be created. Google Artifact Registry and Google Container Registry are both
+excellent choices.
+
+```shell
+# Choose names and locations
+LOCATION="us"
+PROJECT_ID="kpt-demo-73823"
+REPOSITORY_NAME="blueprints"
+
+# Base name for any images in this repository
+REPOSITORY="${LOCATION}-docker.pkg.dev/${PROJECT_ID}/${REPOSITORY_NAME}"
+
+# Create the repository
+gcloud artifacts repositories create --location="${LOCATION}" --repository-format=docker --project="${PROJECT_ID}" "${REPOSITORY_NAME}"
+```
+
+### Creating and pushing a new package
+
+Creating a new package is no different. But when ready to publish, the `kpt pkg push` command 
+is used instead of source control operations.
+
+```shell
+# A package in a new directory
+mkdir hello-world
+kpt pkg init hello-world --description="A simple blueprint"
+
+# Store the contents in the repository, tagged as v1
+kpt pkg push hello-world --image=${REPOSITORY}/hello-world:v1
+
+# The local files are not needed any more, pushing has stored them all
+rm -r hello-world
+```
+
+### Pulling and updating a package
+
+Because the package folder was discarded earlier, the `kpt pkg pull` command 
+is used to place the contents of a particular version at a location. These set of
+commands may be run in `Cloud Build` steps as well, if you are automating the
+publication of packages as part of a CI/CD process.
+
+```shell
+# Recreate the folder and extract the pulled image
+kpt pkg pull hello-world --image=${REPOSITORY}/hello-world:v1
+
+# Add a sub-package from a git repo
+kpt pkg get https://github.com/GoogleCloudPlatform/blueprints.git/catalog/bucket hello-world/my-bucket
+
+# Render to be sure contents are hydrated, and push to a new version tag
+kpt pkg render hello-world
+kpt pkg push hello-world --image=${REPOSITORY}/hello-world:v2
+```
+
+Similar to container image tags, the package image tags like `:v1` and `:v2` above may be used any
+number of ways based on your preferred workflows. The tag `:latest` is used by default if all
+pulls and pushes should read from and overwrite the same location. Semantic tags like `:draft` and
+environmental tags like `:dev`, `:qa`, and `:prod` may be also used.
+
+No matter what tags are used, the image repository and `kpt` cli will treat them as an alphanumeric 
+label.
+
+### Using OCI repository as an upstream
+
+In addition to providing storage for packages, an OCI registry may also 
+be used as a source of upstream images to clone. The `oci://` prefix on this
+command is required to ensure 
+
+```shell
+# Clone the hello-world v1 blueprint into a new folder
+kpt pkg get oci://${REPOSITORY}/hello-world:v1 greetings-planet
+
+# Push the results to the repository, using default `latest` tag in this example
+kpt pkg push greetings-planet --image=${REPOSITORY}/greetings-planet
+```
```

Comment:
Which two operations? Pull and push?

Users directly using a git-based workflow through their preferred git provider with packages in arbitrary parts of git repos would continue to use git directly rather than pull/push.

I want to automate config transformations and abstract the storage layer.

To create a downstream clone of a package, I can do kpt pkg get today. But if I don't want to create a clone with a new upstream, I need to fall back to git clone currently.

To browse/introspect a package, I need to pull it locally then kpt fn source, kpt pkg tree, etc.

To apply a package, I need to pull it locally then kpt live apply.

To transform a package, I need to pull it locally, then kpt fn source, edit and/or kpt fn eval, kpt fn sink, kpt fn render (to enforce invariants), then kpt pkg push to write back a new version.

To upgrade/downgrade a package, I need to pull it locally, then kpt pkg update, kpt fn render (to enforce invariants), then kpt pkg push to write back the new version.

To create a new package, I need to kpt pkg init, create some files (perhaps using kpt fn sink), kpt fn render (to enforce invariants), then kpt pkg push to write back the new version.

Reply:
For comparison:
https://github.com/helm/community/blob/main/hips/hip-0006.md
https://carvel.dev/imgpkg/
https://crossplane.io/docs/v1.3/concepts/packages.html#pushing-a-package
https://cloud.google.com/migrate/anthos/docs/migctl-reference#migctl-artifacts-repo

Reply:
I'd also like to see the ability to create a file that includes kpt package, the requisite container images for the software itself and the kpt functions needed to manipulate the config. This allows easy transport of a complete software distribution.

Reply:
imgpkg supports bundles for that.
https://carvel.dev/imgpkg/docs/latest/air-gapped-workflow/

Reply:
CNAB does also.

Reply:
Are you saying: "these tools already do this so we don't need to" or "these tools do this so we should too"?

I'm going to create a separate issue for this and we can move the discussion there.

Reply:
https://github.com/GoogleContainerTools/kpt/issues/2602

Reply:
I was pointing out existing examples to help people understand the issue and possible approaches to the problem.

Reply:
Ah, got it. The nice thing about CaD is we can rewrite the image fields for people too :)

Reply:
https://github.com/GoogleContainerTools/kpt/issues/2577

Reply:
> Which two operations? Pull and push?

I think I now understand the use-cases for `pkg get` (fork), `pkg pull` and `pkg push` better. I captured my mental model in [this comment](https://github.com/GoogleContainerTools/kpt/pull/2589/files#r760605298).

Reply:
sounds good! (resolving thread)